spec/app/ics-028-cross-chain-valset-update/README.md

```diff
@@ -0,0 +1,257 @@
+
+In the rest of the document, we will discuss the aforementioned state transitions in more detail.
+
+## Function Definitions
+
+### Parent blockchain
+
+This subsection will present the functions executed at the parent blockchain.
+
+```golang
+// expresses will to modify the validator set of the baby blockchain;
+func changeValidatorSet(
+  babyChainId: ChainId
+  valSetUpdate: Validator[]
+) {
+  // enqueue to the queue shared with babyChainId
+  Q[babyChainId].enqueue(valSetUpdate)
+}
+```
+
+- Expected precondition
+  - There exists a blockchain with *babyChainId* identifier
+  - All validators from *valSetUpdate* are validators at the parent blockchain
+- Expected postcondition
+  - The valSetUpdate is enqueued to the queue shared with the blockchain with *babyChainId*
+- Error condition
+  - If the precondition is violated
+
+TODO discuss and write a staking logic on endBlock
+Important: By reading the shared queue Q, the parent blockchain is able to conclude which validators could get their stake back.
+
+### Baby blockchain
+
+```golang
+// initialization function
+func init() {
+  UnbondingQueue = new local FIFO queue
+  validatorSetSeqNum = 0
+  dequeueSeqNum = 0
+  lastObserved = nil
+}
+```
+
+
+```golang
+// function used in End-Block method
+// returns all the elements enqueued since the last invocation of this method
+func observeChangesQ() { 
+  // get the content of the shared queue
+  content = Q.read()
+
+  // check whether the last observed element is in content
+  if (!content.has(lastObserved)) {
+    // update the lastObserved to the last element of content if content is not empty
+    if (!content.empty()) lastObserved = content.last()
+
+    // return the entire content
+    return content
+  } else {
+    // set "old" lastObserved
+    oldLastObserved = lastObserved
+
+    // update the lastObserved to the last element of content
+    lastObserved = content.last()
+
+    // return the entire content of content starting from oldLastObserved (excluding oldLastObserved)
+    return content.startFrom(oldLastObserved)
+  }
+}
+```
+
+- Expected precondition
+  - None
+- Expected postcondition
+  - Returns the elements enqueued since the last Q.read() operation was invoked.
+- Error condition
+  - None
+
+```golang
+// End-Block method executed at the end of each block
+func endBlock(block: Block) {
+  // get time
+  time = block.time
+
+  // finish unbonding for mature validator sets
+  while (!UnbondingQueue.isEmpty()) {
+    // peak the first queue entry
+    seqNum, startTime = UnbondingQueue.peak()
+
+    if (startTime + UNBONDING_PERIOD >= time) {
+      // remove from the unbonding queue
+      seqNum, startTime = UnbondingQueue.dequeue()
+
+      // dequeue all elements until seqNum
+      for (i = dequeueSeqNum + 1; i <= seqNum; i++) {
```

Comment:
dequeueSeqNum not defined

Reply:
It was defined in the init.
Now that we have abandoned the queue semantics, this is removed.